keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,663 @@
+-->
+
+### Goals
+- Allow users to dynamically select the set of namespaces considered when using pod
+  affinity/anti-affinity
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+<!--
+### Non-Goals
+
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I am running a SaaS service, the workload for each customer is placed in a separate
+namespace. The workloads requires 1:1 pod to node placement. I want to use pod 
+anti-affinity across all customers namespaces to achieve that.
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The main risk is performance, using namespace selector will make it easier for 
+users to specify affinity constraints across a large number of namespaces.
+
+The initial implementation of pod affinity/anti-affinity suffered from 
+performance challenges, however over the releases 1.17 - 1.20 we significantly
+improved that. 
+
+We currently have integration test and clusterloader benchmarks that evaluate the extreme
+cases of all pods having affinity/anti-affinity constraints to/against each other.
+Those benchmarks show that the scheduler is able to achieve maximum throughput 
+(i.e., the api-server qps limit).
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+ Add `NamespaceSelector` field to `PodAffinityTerm`:
+
+```go
+type PodAffinityTerm struct {
+    // A label query over the set of namespaces for which the term applies to.
+    // The term is applied to the union of the namespaces selected by this field
+    // and the ones listed in the `Namespaces` field.
+	NamespaceSelector *metav1.LabelSelector
```

Comment:
I started wondering about the following:
- anti-affinity is symetric, so we're ensuring that not only the pods we're scheduling satisfy it, but we also ensure that the pods that are already scheduled will have their anti-affinity satisfied
- with the introduction of NamespaceSelector, this relation is no longer static - in other words, by adding a label to a namespace, I can suddenly add new pods that actually should be considered for these relations

In other words, by adding labels to namespaces - I can suddenly violate existing anti-affinity, which up until now were always satisfied.

Did you think about implications of that?

Reply:
I agree, if the labels change, the outcome will certainly change for future pods.

The other problem I see with namespace selector is performance. While for the incoming pod we can pre-calculate the set of namespaces, it might be expensive to do that for existing pods when evaluating their anti-affinity constraints.

This is likely an API anti-pattern, but what are your thought on changing it to regexp on the namespace name?

Reply:
I'm not huge fan of regexp - though you may want to ask API approvers about their opinion too.

What we can do though is use NamespaceSelector, but in the validation phase just accept the "kubernetes.io/metadata.name' label names (or whatever value we will decide for in #2162 - it's not a regexp, but will allow you to define a set of namespaces in a more native way...

Reply:
Actually - you already can do that with `Namespaces` field, so maybe that doesn't make that much sense...

Reply:
I'm -1 on regexp/prefix matching against names... those are not strong boundaries (a regexp/prefix intended to match "$username-" namespaces would match also "$username-$othervalue-" namespaces)

The dynamic nature is not that different from label-based node affinity/anti-affinity... labels on nodes can change and invalidate previously scheduled pods, right?

Reply:
resolving this, lets discuss the api under the other [comment](https://github.com/kubernetes/enhancements/pull/2251#discussion_r555915819)